benchmarks/apache-spark/src/main/scala/org/renaissance/apache/spark/Als.scala

```diff
@@ -90,8 +92,11 @@ class Als extends RenaissanceBenchmark with SparkUtil {
     RenaissanceBenchmark.deleteTempDir(tempDirPath)
   }
 
-  def runIteration(c: Config): Unit = {
+  def runIteration(c: Config): BenchmarkResult = {
     val als = new org.apache.spark.mllib.recommendation.ALS()
     factModel = als.run(ratings)
+    blackHole(factModel)
```

Comment:
I had the same feeling. Since `runIteration` must now return results to be validated, those results should automatically end up in a blackhole that is handled by the harness.

Reply:
I completely agree. In some cases I already deleted the `blackHole` invocation as the object become part of the returned result. However, in several places the validation is not done (as in *als*) as I have very little idea what to actually validate (so that it makes logical sense). In these places, the original `blackHole` remained. 

For this particular benchmark (*als*), I think it was a bug that it was not there before. I doubt JVM would optimize-out the whole call (`als.run`) without it but it is probably safer to have it there.

Reply:
Maybe EmptyResult should also take the blackholed value?
I also doubt that anything would be optimized, but in the interest of consistency we might want to return some computed value.

Reply:
> Maybe EmptyResult should also take the blackholed value?

I do not think so. As I wrote below, each benchmark should have its own `BenchmarkResult` implementation, `EmptyResult` is a temporary solution until we add proper validation. Ideally, once current benchmarks are equipped with proper validation, `EmptyResult` should be removed and never re-introduced.

The idea is to merge this as it would be needed by #81 and allow other authors (that understand the benchmark code better than I do) add the validation to their benchmarks as time permits.

Reply:
Is the idea then to have some sort of standard interface that the benchmark author must implement and return? If so, should the harness invoke `blackHole` on the returned object?

Reply:
> Is the idea then to have some sort of standard interface that the benchmark author must implement and return? If so, should the harness invoke `blackHole` on the returned object?

Yes, benchmark author must implement the `BenchmarkResult` interface. If we can standardize the kind of information that are needed for validation (and associate it with configuration), we can have helper classes that (at least some) benchmark authors can either use or even subclass. For now, the idea is to get the validation code in place so that we can start running analyses. The rest of refactoring due to #81 should not really change the runtime much, just make the pieces stick together better.

Implementation wise, we should require benchmark authors to store the value that is to be validated in the result object, that should prevent the JVM from optimizing it away (if we call the validate() method). If we want to skip validation, the harness should just assign the benchmark result into a black hole.

BTW, I would suggest that `EmptyResult` outputs a warning about being a placeholder for now (if not fail validation).

Reply:
Ok, sg - my main concern is to (in the final solution) avoid the benchmark authors from having to remember to invoke `blackHole`.

Reply:
> BTW, I would suggest that EmptyResult outputs a warning about being a placeholder for now (if not fail validation).

Currently, `EmptyResult.validate()` prints a warning to standard error output.

> For now, the idea is to get the validation code in place so that we can start running analyses. The rest of refactoring due to #81 should not really change the runtime much, just make the pieces stick together better.
>
> Implementation wise, we should require benchmark authors to store the value that is to be validated in the result object, that should prevent the JVM from optimizing it away (if we call the validate() method). If we want to skip validation, the harness should just assign the benchmark result into a black hole.

Agreed.